### PR TITLE
refactor: make provider trait generic in the whole repo

### PR DESCRIPTION
**Pull request by Eikix**
Time spent on this PR: 1 day

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Resolves: #256 

# What is the new behavior?

Made Provider in all the crates logic generic

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

# Other information

### CODE REVIEW

File: crates/core/src/client/api.rs [hunk 1]
```diff
@@ -6,59 +6,58 @@ use reth_rpc_types::{
...
+use starknet::providers::Provider;
 
 use super::errors::EthApiError;
 use crate::models::balance::TokenBalances;
 use crate::models::transaction::StarknetTransactions;
 
 #[async_trait]
-pub trait KakarotEthApi<T: JsonRpcTransport>: KakarotStarknetApi<T> {
-    async fn block_number(&self) -> Result<U64, EthApiError<T::Error>>;
+pub trait KakarotEthApi<P: Provider + Send + Sync>: KakarotStarknetApi<P> + Send + Sync {
```
Comment (danilowhk) [inline crates/core/src/client/api.rs:16]: Why do we need "P: Provider + Send + Sync" and then "<P> + Send + Sync"

Comment (Eikix) [reply crates/core/src/client/api.rs:16]: Good question, somehow we need each one to be able to be passed around threads

File: crates/core/src/client/errors.rs [hunk 2]
```diff
@@ -35,10 +34,10 @@ pub enum ConfigError {
 
 /// Error that can accure when interacting with the Kakarot ETH API.
 #[derive(Debug, Error)]
-pub enum EthApiError<T: std::error::Error> {
+pub enum EthApiError<E: std::error::Error> {
     /// Request to the Starknet provider failed.
     #[error(transparent)]
-    RequestError(#[from] ProviderError<JsonRpcClientError<T>>),
+    RequestError(#[from] ProviderError<E>),
```
Comment (danilowhk) [inline crates/core/src/client/errors.rs:40]: Here we had JsonRpcClientError<T> before, when using JsonRpc and getting and error, would this format work?

File: crates/core/src/client/mod.rs [hunk 3]
```diff
@@ -44,13 +43,13 @@ use crate::models::felt::Felt252Wrapper;
 use crate::models::transaction::{StarknetTransaction, StarknetTransactions};
 use crate::models::ConversionError;
 
-pub struct KakarotClient<T: Provider> {
-    starknet_provider: T,
+pub struct KakarotClient<P: Provider + Send + Sync> {
```
Comment (danilowhk) [inline crates/core/src/client/mod.rs:46]: Do we need Send + Sync?

File: crates/core/src/client/errors.rs [hunk 4]
```diff
@@ -92,12 +91,21 @@ impl<T: std::error::Error> From<EthApiError<T>> for ErrorObject<'static> {
...
+impl<E: std::error::Error> From<EthApiError<E>> for jsonrpsee::core::Error {
+    fn from(err: EthApiError<E>) -> Self {
         jsonrpsee::core::Error::Call(err.into())
     }
 }
 
+/// Converting directly from `ConversionError` to `jsonrpsee::core::Error` in case conversions are
+/// done in eth-rpc crate.
+/// TODO: Remove this when conversions are done in core crate only.
+impl From<ConversionError> for ErrorObject<'static> {
```
Comment (greged93) [inline crates/core/src/client/errors.rs:103]: I think we can remove this and just use `into()` to go from ConversionError to EthApiError. The idea of conversion error is that it should always live inside a EthApiError, hence the from implementation.

**Review by greged93 — COMMENTED** [review]
One smol comment

File: crates/core/src/client/errors.rs (hunk 2 above)
Comment (Eikix) [reply crates/core/src/client/errors.rs:40]: We needed something more modular

File: crates/core/src/client/errors.rs (hunk 4 above)
Comment (Eikix) [reply crates/core/src/client/errors.rs:103]: alright!

**Review by greged93 — APPROVED** [review]
Lgtm!

**Pull request merged**